packages/eslint-plugin-react-hooks/__tests__/ESLintRuleReactiveDeps-test.js

```diff
@@ -0,0 +1,1442 @@
+      errors: [
+        "React Hook useEffect has a missing dependency: 'local'. " +
+          'Either include it or remove the dependency array.',
+      ],
+    },
+    {
+      code: `
+        // Regression test
+        function MyComponent() {
+          const local = 42;
+          useEffect(() => {
+            function inner() {
+              console.log(local);
+            }
+            inner();
+          }, []);
+        }
+      `,
+      output: `
+        // Regression test
+        function MyComponent() {
+          const local = 42;
+          useEffect(() => {
+            function inner() {
+              console.log(local);
+            }
+            inner();
+          }, [local]);
+        }
+      `,
+      errors: [
+        "React Hook useEffect has a missing dependency: 'local'. " +
+          'Either include it or remove the dependency array.',
+      ],
+    },
+    {
+      code: `
+        function MyComponent() {
+          const local1 = 42;
+          {
+            const local2 = 42;
+            useEffect(() => {
+              console.log(local1);
+              console.log(local2);
+            }, []);
+          }
+        }
+      `,
+      output: `
+        function MyComponent() {
+          const local1 = 42;
+          {
+            const local2 = 42;
+            useEffect(() => {
+              console.log(local1);
+              console.log(local2);
+            }, [local1, local2]);
+          }
+        }
+      `,
+      errors: [
+        "React Hook useEffect has missing dependencies: 'local1' and 'local2'. " +
+          'Either include them or remove the dependency array.',
+      ],
+    },
+    {
+      code: `
+        function MyComponent() {
+          const local1 = 42;
+          const local2 = 42;
+          useEffect(() => {
+            console.log(local1);
+            console.log(local2);
+          }, [local1]);
+        }
+      `,
+      output: `
+        function MyComponent() {
+          const local1 = 42;
+          const local2 = 42;
+          useEffect(() => {
+            console.log(local1);
+            console.log(local2);
+          }, [local1, local2]);
+        }
+      `,
+      errors: [
+        "React Hook useEffect has a missing dependency: 'local2'. " +
+          'Either include it or remove the dependency array.',
+      ],
+    },
+    {
+      code: `
+        function MyComponent() {
+          const local1 = 42;
+          const local2 = 42;
+          useEffect(() => {
+            console.log(local1);
+          }, [local1, local2]);
+        }
+      `,
+      output: `
+        function MyComponent() {
+          const local1 = 42;
+          const local2 = 42;
+          useEffect(() => {
+            console.log(local1);
+          }, [local1]);
+        }
+      `,
+      errors: [
+        "React Hook useEffect has an unnecessary dependency: 'local2'. " +
+          'Either exclude it or remove the dependency array.',
+      ],
```

Comment:
This kinda runs into that issue I mentioned in an earlier comment. It may be valid to use the latest value of a local but only reacting to a specific value changing. I know I wrote something that used this property recently but I can't remember what...

At any rate, it would be good to opt out of specifying a specific value without disabling the rule for the entire array. Perhaps it could be possible to write something to let the rule know you know about it? My first idea would be just having the identifier present inside the array but commented out, but I'm not sure eslint lets you consume comments that easily

Reply:
Can you give me a specific example?

Reply:
Here's some partial source code where I ended up having to break this rule. You can see it in the last `useEffect`'s effect list -- basically, I want this effect to run after the update is called by a dispatch to `setCurrentState`. However, it should not run on initial mount, or when `setCurrentState` was invoked by the previous effect which synchronizes the internal-only `page` with the `memoPage`.

```ts
function useCurrentPage({
  location,
  marker,
  replace
}: Pick<Props, 'location' | 'marker' | 'replace'>) {
  const { hash } = location
  const memoPage = useMemo(() => {
    if (!hash || !hash.startsWith('#')) {
      return 1
    }
    return Math.max(
      1,
      parseInt(hash.slice(1), 10) || getPageFromChapterId(hash) || 1
    )
  }, [hash])
  const [page, setCurrentPage] = useState(memoPage)

  useEffect(() => {
    setCurrentPage(memoPage)
  }, [memoPage])

  useEffect(() => {
    // ignore if it was an update from the synchronization above
    if (page === memoPage) {
      return
    }

    if ((!page || page <= 1) && marker === null) {
      replace({
        ...location,
        hash: ''
      })
    } else {
      replace({
        ...location,
        hash: `#${page}`
      })
    }
    // uses latest data but only triggered by setCurrentPage changes
  }, [page])

  return [memoPage, setCurrentPage] as [typeof memoPage, typeof setCurrentPage]
}
```

It's certainly possible I just didn't think this through well enough, but it looks unavoidable.

Reply:
The rubber ducky debugging might have struck again, because it looks like the `page === memoPage` check in the effect is also sufficient to guard against unwanted updates to `marker` or `location`. However, it'd still be a problem to have `memoPage` in the dependencies list -- the effect would run before the synchronization, used to detect whether `setCurrentPage` was dispatched from inside or from outside, happens.

Reply:
I think this use case I've got also relates to this:

```jsx
function LinkInterceptor({ html }) {
  const handleClick = e => {
    e.preventDefault()
    console.log(e.target.href)
  }
  const ref = useRef()

  useEffect(
    () => {
      Array.from(ref.current.querySelectorAll('a')).forEach(node =>
        node.addEventListener('click', handleClick)
      )

      return () => {
        Array.from(ref.current.querySelectorAll('a')).forEach(node =>
          node.removeEventListener('click', handleClick)
        )
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [html]
  )

  return <div ref={ref} dangerouslySetInnerHTML={{ __html: html }} />
}
```

Reply:
Please post both of these in https://github.com/facebook/react/issues/14920, with the guidelines I mentioned. Thanks.